README.md
# Pocker
A doc project


CHANGELOG.md
<a name="1.0.0"></a>
# 1.0.0 (2017-09-19)


### Bug Fixes

* 123213 ([9b78789](https://github.com/blankPen/Pocker/commit/9b78789))
* **213:** 123213 ([cdbcd54](https://github.com/blankPen/Pocker/commit/cdbcd54))





script/API.js
/**
 * 创建一篇文章
 */
require('./color');
const fs = require('fs');
const path = require('path');
const moment = require('moment');

const postPath = path.resolve(process.cwd(), 'post');
const sidePath = path.resolve(process.cwd(), 'post/_sidebar.md');
const getFileName = (name, now) => `${now.format('YYYY-MM-DD')}_${name}.md`;
const postTemp = (name, now) => `
# ${name}
------
> Created on ${now.format('YYYY-MM-DD hh:mm')};
`;

function rewriteSidebar(postList) {
  let sbContent = fs.readFileSync(sidePath).toString('utf8');
  let newSb = '- [首页](/)';
  postList.forEach(item => {
    newSb += `\n- [${item.name}](${item.filename})`
  });
  fs.writeFileSync(sidePath, newSb);
}

function createPost(name) {
  const now = moment();
  let postList = getList();
  let fileNames = postList.map(({ filename }) => filename);
  let filename = getFileName(name, now);
  if (fileNames.indexOf(filename) !== -1) {
    return `文章 ${filename} 已存在`;
  }

  try {
    fs.writeFileSync(path.resolve(postPath, filename), postTemp(name, now));
    postList.push({
      name: name,
      filename: filename
    });
    rewriteSidebar(postList);
    console.log('创建成功'.green);
  } catch (error) {
    return error.message;
  }
}

function deletePost(filename) {
  let postList = getList();
  let fileNames = postList.map(({ filename }) => filename);
  const index = fileNames.indexOf(filename);
  fs.unlinkSync(path.resolve(postPath, filename));
  postList = [
    ...postList.slice(0, index),
    ...postList.slice(index + 1, postList.length),
  ];
  rewriteSidebar(postList);
  console.log('删除成功'.green);
}

function getList() {
  let sbContent = fs.readFileSync(sidePath).toString('utf8');
  let postList = sbContent.split('\n').map(str => {
    var n = str.match(/\[.+\]/)[0];
    var p = str.match(/\(.+\)/)[0];
    return {
      name: n.substring(1, n.length - 1),
      filename: p.substring(1, p.length - 1),
    };
  });
  postList.shift();
  return postList;
}

module.exports = {
  add: createPost,
  delete: deletePost,
  getList: getList,
}

script/post.js
/**
 * 创建一篇文章
 */
require('./color');
const repl = require('./prompt');
const fs = require('fs');
const path = require('path');
const moment = require('moment');
const { argv } = require('optimist');

const postPath = path.resolve(process.cwd(), 'docs');
const sidePath = path.resolve(process.cwd(), 'docs/sidebar.md');
let sbContent = fs.readFileSync(sidePath).toString('utf8');
let postList = sbContent.split('\n').map(str => {
  var n = str.match(/\[.+\]/)[0];
  var p = str.match(/\(.+\)/)[0];
  return {
    name: n.substring(1, n.length - 1),
    filename: p.substring(1, p.length - 1),
  };
});
postList.shift();
let fileNames = postList.map(({ filename }) => filename);



const now = moment();

const getFileName = name => `${now.format('YYYY-MM-DD')}_${name}.md`;
const postTemp = name => `
  # ${name}
  ------
  > Created on ${now.format('YYYY-MM-DD hh:mm')};
`;

function rewriteSidebar(name) {
  let newSb = '- [首页](/)';
  postList.forEach(item => {
    newSb += `\n- [${item.name}](${item.filename})`
  });
  fs.writeFileSync(sidePath, newSb);
}

function createPost(name) {

  try {
    fs.writeFile(path.resolve(postPath, getFileName(name)), postTemp(name), () => {
      console.log('创建成功'.green);
      postList.push({
        name: name,
        filename: getFileName(name)
      });
      rewriteSidebar();
    });
  } catch (error) {
    console.error(error);
  }
}
function deletePost(filename) {
  const index = fileNames.indexOf(filename);
  fs.unlinkSync(path.resolve(postPath, filename));
  postList = [
    ...postList.slice(0, index),
    ...postList.slice(index + 1, postList.length),
  ];
  rewriteSidebar();
}

const createPrompt = [
  {
    name: 'name',
    title: '请输入文章名称',
    default: '',
    validate: (value) => {
      if (!value) {
        return '文章名不能为空';
      }
      filename = getFileName(value);
      if (fileNames.indexOf(filename) !== -1) {
        return `文章 ${filename} 已存在`;
      }
      return true;
    },
  },
  {
    name: 'confirm',
    title: (data) => {
      return `创建文章 ${data.name} [Y/n]`;
    },
    default: 'y',
  },
];

const deletePrompt = [
  {
    name: 'name',
    title: '请选择要删除的文章',
    options: fileNames,
  }, {
    name: 'confirm',
    title: (data) => {
      return `删除文章 ${data.name} [Y/n]`;
    },
    default: 'y',
  },
];

const type = argv.type;
if (type === 'c' || type === 'create') {
  repl(createPrompt, (data) => {
    if (/y/i.test(data.confirm)) {
      createPost(data.name);
    }
  });
} else if (type === 'd' || type === 'delete') {
  if(!fileNames.length){
    return console.log('没有可删除的文章\n'.red)
  }
  repl(deletePrompt, (data) => {
    if (/y/i.test(data.confirm)) {
      deletePost(data.name);
    }
  });
}

script/prompt.js
require('./color');

const { stdin, stdout } = process;

const getDesc = (desc, data) => {
  if (typeof desc === 'function') {
    return desc(data);
  }
  return `${desc}`;
};

const wait = (q, data) => {
  const title = getDesc(q.title, data);
  const description = getDesc(q.description, data);

  stdout.write(`${'?'.green} ${title.white}\n`);
  if (q.description) {
    stdout.write(`  ${description.grey}\n: `);
  }
  if (Array.isArray(q.options)) {
    q.options.forEach((op, i) => {
      stdout.write(`${`${i + 1})`.cyan} ${op.grey}\n`);
    });
    stdout.write('\n: ');
  }
};

module.exports = (questions, callback) => {
  const result = {};

  let [q] = questions;
  let i = 0;

  wait(q, result);
  stdin.resume();
  stdin.setEncoding('utf8');

  stdin.on('data', (chunk) => {
    chunk = chunk.replace(/[\s\n]/, '');

    result[q.name] = chunk || q.default;

    if (Array.isArray(q.options)) {
      result[q.name] = q.options[chunk - 1];
    }

    const { validate } = questions[i];

    if (typeof validate === 'function') {
      const val = validate(chunk);
      if (val !== true) {
        stdout.write(`✘ ${val}，请重新输入\n`.red);
        i -= 1;
      }
    }

    i += 1;
    q = questions[i];

    if (i >= questions.length) {
      stdin.pause();
      callback(result);
    } else {
      wait(q, result);
    }
  });
};


script/server.js
'use strict'

require('./color.js');
const serveStatic = require('serve-static');
const serveIndex = require('serve-index');
const express = require('express');
const bodyParser = require('body-parser');
const livereload = require('connect-livereload');
const lrserver = require('livereload');
const path = require('path');
const { argv } = require('optimist');

const API = require('./API.js');

const doSuccess = (res, data = true) => {
  res.writeHead(200, { 'Content-Type': 'application/json;charset=UTF-8' });
  res.end(JSON.stringify({
    data,
  }));
}
const do404 = res => {
  res.writeHead(404, { 'Content-Type': 'application/json;charset=UTF-8' });
  res.end();
}


function createServer(filePath = '.', port = 3000, livereloadPort = 35729) {
  let cwd = path.resolve(filePath)
  let app = express();
  app.use(bodyParser.json());
  app.use(bodyParser.urlencoded({ extended: true }));

  app.get('/api/setting/post/list', (req, res) => {
    res.send({
      data: API.getList(),
    });
  });
  app.post('/api/setting/post/add', (req, res) => {
    let err = API.add(req.body.name);
    res.send({
      err,
    });
  });
  app.post('/api/setting/post/delete', (req, res) => {
    let err = API.delete(req.body.name);
    res.send({
      err,
    });
  });

  app.use(livereload())
  app.use(serveStatic(cwd))
  app.use(serveIndex(cwd, { icons: true }));
  app.listen(port)
  lrserver.createServer({
    exts: ['md'],
    exclusions: ['node_modules/'],
    port: livereloadPort
  }).watch(cwd)



  console.log(`\nListening at `.green + `http://localhost:${port}`.yellow + '\n');
}
createServer('.', argv.p, 35729);

script/pre-commit.js

/**
 * Git COMMIT-MSG hook for validating commit message
 * See https://docs.google.com/document/d/1rk04jEuGfk9kYzfqCuOlPTSJw3hEDZJTBN5E5f1SALo/edit
 *
 * Installation:
 * >> cd <angular-repo>
 * >> ln -s ../../validate-commit-msg.js .git/hooks/commit-msg
 */
/* eslint-disable */

'use strict';

var fs = require('fs');
var util = require('util');
var resolve = require('path').resolve;
var findup = require('findup');
var semverRegex = require('semver-regex');

var config = getConfig();
var MAX_LENGTH = config.maxSubjectLength || 100;
var IGNORED = new RegExp(util.format('(^WIP)|(^%s$)', semverRegex().source));

// fixup! and squash! are part of Git, commits tagged with them are not intended to be merged, cf. https://git-scm.com/docs/git-commit
var PATTERN = /^((fixup! |squash! )?(\w+)(?:\(([^\)\s]+)\))?: (.+))(?:\n|$)/;
var MERGE_COMMIT_PATTERN = /^Merge /;
var error = function () {
  // gitx does not display it
  // http://gitx.lighthouseapp.com/projects/17830/tickets/294-feature-display-hook-error-message-when-hook-fails
  // https://groups.google.com/group/gitx/browse_thread/thread/a03bcab60844b812
  console[config.warnOnFail
    ? 'warn'
    : 'error']('INVALID COMMIT MSG: ' + util.format.apply(null, arguments));
};

var validateMessage = function (raw) {
  var types = ['feat',
    'fix',
    'docs',
    'style',
    'refactor',
    'perf',
    'test',
    'build',
    'ci',
    'chore',
    'revert'];

  var messageWithBody = (raw || '').split('\n').filter(function (str) {
    return str.indexOf('#') !== 0;
  }).join('\n');

  var message = messageWithBody.split('\n').shift();

  if (message === '') {
    console.log('Aborting commit due to empty commit message.');
    return false;
  }

  var isValid = true;

  if (MERGE_COMMIT_PATTERN.test(message)) {
    console.log('Merge commit detected.');
    return true
  }

  if (IGNORED.test(message)) {
    console.log('Commit message validation ignored.');
    return true;
  }

  var match = PATTERN.exec(message);

  if (!match) {
    error('提交信息格式不通过，具体格式请按照："<type>(<scope>): <subject>" !');
    isValid = false;
  } else {
    var firstLine = match[1];
    var squashing = !!match[2];
    var type = match[3];
    var scope = match[4];
    var subject = match[5];

    var SUBJECT_PATTERN = new RegExp(config.subjectPattern || '.+');
    var SUBJECT_PATTERN_ERROR_MSG = config.subjectPatternErrorMsg || 'subject does not match subject pattern!';

    if (firstLine.length > MAX_LENGTH && !squashing) {
      error('is longer than %d characters !', MAX_LENGTH);
      isValid = false;
    }

    if (types !== '*' && types.indexOf(type) === -1) {
      error('"%s" is not allowed type ! Valid types are: %s', type, types.join(', '));
      isValid = false;
    }

    if (!SUBJECT_PATTERN.exec(subject)) {
      error(SUBJECT_PATTERN_ERROR_MSG);
      isValid = false;
    }
  }

  // Some more ideas, do want anything like this ?
  // - Validate the rest of the message (body, footer, BREAKING CHANGE annotations)
  // - allow only specific scopes (eg. fix(docs) should not be allowed ?
  // - auto correct the type to lower case ?
  // - auto correct first letter of the subject to lower case ?
  // - auto add empty line after subject ?
  // - auto remove empty () ?
  // - auto correct typos in type ?
  // - store incorrect messages, so that we can learn

  isValid = isValid || config.warnOnFail;

  if (isValid) { // exit early and skip messaging logics
    return true;
  }

  var argInHelp = config.helpMessage && config.helpMessage.indexOf('%s') !== -1;

  if (argInHelp) {
    console.log(config.helpMessage, messageWithBody);
  } else if (message) {
    console.log(message);
  }

  if (!argInHelp && config.helpMessage) {
    console.log(config.helpMessage);
  }

  return false;
};

// publish for testing
exports.validateMessage = validateMessage;
exports.getGitFolder = getGitFolder;
exports.config = config;

// hacky start if not run by mocha :-D
// istanbul ignore next
if (process.argv.join('').indexOf('mocha') === -1) {

  var commitMsgFile = process.argv[2] || getGitFolder() + '/COMMIT_EDITMSG';
  var incorrectLogFile = commitMsgFile.replace('COMMIT_EDITMSG', 'logs/incorrect-commit-msgs');

  var hasToString = function hasToString(x) {
    return x && typeof x.toString === 'function';
  };
  fs.readFile(commitMsgFile, function (err, buffer) {
    var msg = getCommitMessage(buffer);
    if (!validateMessage(msg)) {
      fs.appendFile(incorrectLogFile, msg + '\n', function () {
        process.exit(1);
      });
    } else {
      process.exit(0);
    }

    function getCommitMessage(buffer) {
      return hasToString(buffer) && buffer.toString();
    }
  });
}

function getConfig() {
  var pkgFile = findup.sync(process.cwd(), 'package.json');
  var pkg = JSON.parse(fs.readFileSync(resolve(pkgFile, 'package.json')));
  return pkg && pkg.config && pkg.config['validate-commit-msg'] || {};
}

function getGitFolder() {
  var gitDirLocation = './.git';
  if (!fs.existsSync(gitDirLocation)) {
    throw new Error('Cannot find file ' + gitDirLocation);
  }

  if (!fs.lstatSync(gitDirLocation).isDirectory()) {
    var unparsedText = '' + fs.readFileSync(gitDirLocation);
    gitDirLocation = unparsedText.substring('gitdir: '.length).trim();
  }

  if (!fs.existsSync(gitDirLocation)) {
    throw new Error('Cannot find file ' + gitDirLocation);
  }

  return gitDirLocation;
}


script/validate-commit-msg.js

/**
 * Git COMMIT-MSG hook for validating commit message
 * See https://docs.google.com/document/d/1rk04jEuGfk9kYzfqCuOlPTSJw3hEDZJTBN5E5f1SALo/edit
 *
 * Installation:
 * >> cd <angular-repo>
 * >> ln -s ../../validate-commit-msg.js .git/hooks/commit-msg
 */
/* eslint-disable */

'use strict';

var fs = require('fs');
var util = require('util');
var resolve = require('path').resolve;
var findup = require('findup');
var semverRegex = require('semver-regex');

var config = getConfig();
var MAX_LENGTH = config.maxSubjectLength || 100;
var IGNORED = new RegExp(util.format('(^WIP)|(^%s$)', semverRegex().source));

// fixup! and squash! are part of Git, commits tagged with them are not intended to be merged, cf. https://git-scm.com/docs/git-commit
var PATTERN = /^((fixup! |squash! )?(\w+)(?:\(([^\)\s]+)\))?: (.+))(?:\n|$)/;
var MERGE_COMMIT_PATTERN = /^Merge /;
var error = function () {
  // gitx does not display it
  // http://gitx.lighthouseapp.com/projects/17830/tickets/294-feature-display-hook-error-message-when-hook-fails
  // https://groups.google.com/group/gitx/browse_thread/thread/a03bcab60844b812
  console[config.warnOnFail
    ? 'warn'
    : 'error']('INVALID COMMIT MSG: ' + util.format.apply(null, arguments));
};

var validateMessage = function (raw) {
  var types = ['feat',
    'fix',
    'docs',
    'style',
    'refactor',
    'perf',
    'test',
    'build',
    'ci',
    'chore',
    'revert'];

  var messageWithBody = (raw || '').split('\n').filter(function (str) {
    return str.indexOf('#') !== 0;
  }).join('\n');

  var message = messageWithBody.split('\n').shift();

  if (message === '') {
    console.log('Aborting commit due to empty commit message.');
    return false;
  }

  var isValid = true;

  if (MERGE_COMMIT_PATTERN.test(message)) {
    console.log('Merge commit detected.');
    return true
  }

  if (IGNORED.test(message)) {
    console.log('Commit message validation ignored.');
    return true;
  }

  var match = PATTERN.exec(message);

  if (!match) {
    error('提交信息格式不通过，具体格式请按照："<type>(<scope>): <subject>" !');
    isValid = false;
  } else {
    var firstLine = match[1];
    var squashing = !!match[2];
    var type = match[3];
    var scope = match[4];
    var subject = match[5];

    var SUBJECT_PATTERN = new RegExp(config.subjectPattern || '.+');
    var SUBJECT_PATTERN_ERROR_MSG = config.subjectPatternErrorMsg || 'subject does not match subject pattern!';

    if (firstLine.length > MAX_LENGTH && !squashing) {
      error('is longer than %d characters !', MAX_LENGTH);
      isValid = false;
    }

    if (types !== '*' && types.indexOf(type) === -1) {
      error('"%s" is not allowed type ! Valid types are: %s', type, types.join(', '));
      isValid = false;
    }

    if (!SUBJECT_PATTERN.exec(subject)) {
      error(SUBJECT_PATTERN_ERROR_MSG);
      isValid = false;
    }
  }

  // Some more ideas, do want anything like this ?
  // - Validate the rest of the message (body, footer, BREAKING CHANGE annotations)
  // - allow only specific scopes (eg. fix(docs) should not be allowed ?
  // - auto correct the type to lower case ?
  // - auto correct first letter of the subject to lower case ?
  // - auto add empty line after subject ?
  // - auto remove empty () ?
  // - auto correct typos in type ?
  // - store incorrect messages, so that we can learn

  isValid = isValid || config.warnOnFail;

  if (isValid) { // exit early and skip messaging logics
    return true;
  }

  var argInHelp = config.helpMessage && config.helpMessage.indexOf('%s') !== -1;

  if (argInHelp) {
    console.log(config.helpMessage, messageWithBody);
  } else if (message) {
    console.log(message);
  }

  if (!argInHelp && config.helpMessage) {
    console.log(config.helpMessage);
  }

  return false;
};

// publish for testing
exports.validateMessage = validateMessage;
exports.getGitFolder = getGitFolder;
exports.config = config;

// hacky start if not run by mocha :-D
// istanbul ignore next
if (process.argv.join('').indexOf('mocha') === -1) {

  var commitMsgFile = process.argv[2] || getGitFolder() + '/COMMIT_EDITMSG';
  var incorrectLogFile = commitMsgFile.replace('COMMIT_EDITMSG', 'logs/incorrect-commit-msgs');

  var hasToString = function hasToString(x) {
    return x && typeof x.toString === 'function';
  };

  fs.readFile(commitMsgFile, function (err, buffer) {
    var msg = getCommitMessage(buffer);

    if (!validateMessage(msg)) {
      fs.appendFile(incorrectLogFile, msg + '\n', function () {
        process.exit(1);
      });
    } else {
      process.exit(0);
    }

    function getCommitMessage(buffer) {
      return hasToString(buffer) && buffer.toString();
    }
  });
}

function getConfig() {
  var pkgFile = findup.sync(process.cwd(), 'package.json');
  var pkg = JSON.parse(fs.readFileSync(resolve(pkgFile, 'package.json')));
  return pkg && pkg.config && pkg.config['validate-commit-msg'] || {};
}

function getGitFolder() {
  var gitDirLocation = './.git';
  if (!fs.existsSync(gitDirLocation)) {
    throw new Error('Cannot find file ' + gitDirLocation);
  }

  if (!fs.lstatSync(gitDirLocation).isDirectory()) {
    var unparsedText = '' + fs.readFileSync(gitDirLocation);
    gitDirLocation = unparsedText.substring('gitdir: '.length).trim();
  }

  if (!fs.existsSync(gitDirLocation)) {
    throw new Error('Cannot find file ' + gitDirLocation);
  }

  return gitDirLocation;
}
